### PR TITLE
Added support for aten::randperm and aten::polar

### PR DESCRIPTION
**Pull request by vijaykr338**
### Details:
 added support for aten::randperm and tests

 currently working on aten::polar and others 

### Tickets:
 - #29547

### CODE REVIEW

**Review by vijaykr338 — COMMENTED** [review]
I found out that pytorch.polar returns complex tensors meanwhile OpenVINO does not support them right now

to make it work, I separately computed real and imag parts using cos and sine then wrapped it with a ComplexTypeMark.

File: src/frontends/pytorch/src/op/polar.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+OutputVector translate_polar(const NodeContext& context) {
+    num_inputs_check(context, 2, 3);
+    auto abs = context.get_input(0);
+    auto angle = context.get_input(1);
+    auto real = context.mark_node(std::make_shared<v1::Multiply>(abs,context.mark_node(std::make_shared<v0::Cos>(angle))));
+    auto imag = context.mark_node(std::make_shared<v1::Multiply>(abs,context.mark_node(std::make_shared<v0::Sin>(angle))));
+    auto complex_concat = context.mark_node(std::make_shared<v0::Concat>(OutputVector{real, imag}, -1));
+    // wrap the tensor with ComplexTypeMark to flag it as complex for later operations.
+    auto complex_tensor = context.mark_node(std::make_shared<ComplexTypeMark>(complex_concat));
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/polar.cpp:24]:
```suggestion
    // wrap the tensor with ComplexTypeMark to flag it as complex for later operations.
    auto complex_tensor = context.mark_node(std::make_shared<ComplexTypeMark>(real, imag));
```
Instead of concat you can pass the real and imaginary parts to complex mark.

File: tests/layer_tests/pytorch_tests/test_randperm.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        compiled_model = core.compile_model(converted_model, ie_device).
+        ov_output_dict = compiled_model(())
+        ov_output_tensor = list(ov_output_dict.values())[0]
+
+        assert ov_output_tensor.shape[0] == n, f"Output shape {ov_output_tensor.shape} does not match expected ({n},)"
+        assert self.is_valid_permutation(ov_output_tensor, n), (
+            f"Output {ov_output_tensor} is not a valid permutation of [0, 1, ..., {n-1}]"
+        )
+
+    @pytest.mark.xfail(reason="OpenVINO doesn't support empty tensors for randperm")
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_randperm.py:62]: Openvino can create an empty tensor:

```
auto const_empty = std::make_shared<v0::Constant>(element::i64, Shape{0}, std::vector<int64_t>{});
```

**Review by mvafin — CHANGES_REQUESTED** [review]

File: tests/layer_tests/pytorch_tests/test_polar.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            np.array([1.0, 2.0, 3.0], dtype=np.float32),
+            np.array([0.1, 0.2, 0.3], dtype=np.float32)
+        )
+
+    def create_model(self):
+        class PolarModel(torch.nn.Module):
+            def forward(self, abs, angle):
+                real = abs * torch.cos(angle)
+                imag = abs * torch.sin(angle)
+                return torch.stack([real, imag], dim=-1)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_polar.py:21]: You are not using aten::polar here. Please use it and return the value using `torch.view_as_real`

**Comment by vijaykr338** [comment]
@mvafin, 

Thank you very much for your review, I added the suggested changes along with the coding style this time

File: src/frontends/pytorch/src/op_table.cpp [hunk 4]
```diff
@@ -635,12 +637,14 @@ const std::unordered_map<std::string, CreatorFunction> get_supported_ops_ts() {
...
         {"aten::pixel_unshuffle", op::translate_pixel_unshuffle},
         {"aten::prelu", op::translate_1to1_match_2_inputs<opset10::PRelu>},
+        {"aten::polar", op::translate_polar},
         {"aten::pow", op::translate_pow},
         {"aten::pow_", op::translate_pow},
         {"aten::prod", op::translate_prod},
         {"aten::quantize_per_channel", op::translate_quantize_per_channel},
         {"aten::quantize_per_tensor", op::translate_quantize_per_tensor},
         {"aten::rand", op::translate_rand},
+        {"aten::rand", op::translate_randperm},
```
Comment (mvafin) [inline src/frontends/pytorch/src/op_table.cpp:647]:
```suggestion
        {"aten::randperm", op::translate_randperm},
```

**Review by mvafin — CHANGES_REQUESTED** [review]
Tests for `aten::randperm` fail

**Comment by vijaykr338** [comment]
@mvafin I have updated the test for aten::polar to use kwargs_to_prepare_input instead of lambda 

the tests run fine for on my laptop for randperm

since randperm produces random permutation, we can't expect the output of OpenVINO to match Pytorch's output, so instead we just verify if the output has correct shape and when sorted gives 0 to n-1 

therefore, ensuring that the permutation is vaild, even if the order is different

File: src/frontends/pytorch/src/op/polar.cpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+using namespace ov::op;
+
+OutputVector translate_polar(const NodeContext& context) {
+    num_inputs_check(context, 2, 3);
+    auto abs = context.get_input(0);
+    auto angle = context.get_input(1);
+    auto real =
+        context.mark_node(std::make_shared<v1::Multiply>(abs, context.mark_node(std::make_shared<v0::Cos>(angle))));
+    auto imag =
+        context.mark_node(std::make_shared<v1::Multiply>(abs, context.mark_node(std::make_shared<v0::Sin>(angle))));
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/polar.cpp:22]: Please make `Cos` and `Sin` on a separate lines for easier reading

File: src/frontends/pytorch/src/op/randperm.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        if (!context.input_is_none(1)) {
+            dtype_value = context.const_input<int>(1);
+            OPENVINO_ASSERT(dtype_value == 4,
+                            "Only dtype value 4 (int64) is supported for aten::randperm, got: ",
+                            dtype_value);
+        }
+    } else if (num_inputs == 5) {
+        if (!context.input_is_none(1)) {
+            dtype_value = context.const_input<int>(1);
+            OPENVINO_ASSERT(dtype_value == 4,
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/randperm.cpp:30]: Instead of `OPENVINO_ASSERT`, please use `PYTORCH_OP_CONVERSION_CHECK`

Comment (vijaykr338) [reply src/frontends/pytorch/src/op/randperm.cpp]: understood

File: tests/layer_tests/pytorch_tests/test_randperm.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                if self.num_inputs == 1:
+                    p = torch.randperm(self.n)
+                elif self.num_inputs == 2:
+                    p = torch.randperm(self.n, dtype=self.dtype)
+                elif self.num_inputs == 5:
+                    p = torch.randperm(self.n, dtype=self.dtype, layout=torch.strided, 
+                                         device=x.device, pin_memory=False)
+                else:
+                    raise ValueError("Invalid num_inputs")
+                sorted_p, _ = torch.sort(p)
```
[the remark this answers is not in the document]
Comment (vijaykr338) [reply tests/layer_tests/pytorch_tests/test_randperm.py]: alright, I updated it so that randperm's o/p shuffles the actual tensor, and then sorts it 

please check

**Comment by vijaykr338** [comment]
@mvafin 
should I make any further changes?

File: src/frontends/pytorch/src/op/randperm.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+namespace ov {
+namespace frontend {
+namespace pytorch {
+namespace op {
+
+using namespace ov::op;
+
+OutputVector translate_randperm(const NodeContext& context) {
+    auto num_inputs = context.get_input_size();
+    int64_t n = context.const_input<int64_t>(0);
```
[the remark this answers is not in the document]
Comment (mvafin) [reply src/frontends/pytorch/src/op/randperm.cpp]: What error is printed in the n == 0 case? How Squeeze can help?

File: tests/layer_tests/pytorch_tests/test_polar.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    ])
+    @pytest.mark.parametrize("dtype", [
+        np.float32,
+        np.float64
+    ])
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    def test_polar(self, input_case, dtype, ie_device, precision, ir_version):
+        atol = 1e-4 if precision == "FP32" else 1e-3
+        rtol = 1e-4
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_polar.py:39]:
```suggestion
```
you are not using `atol` and `rtol` here

File: tests/layer_tests/pytorch_tests/test_randperm.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                    p = torch.randperm(self.n)
+                elif self.num_inputs == 2:
+                    p = torch.randperm(self.n, dtype=self.dtype)
+                elif self.num_inputs == 5:
+                    p = torch.randperm(self.n, dtype=self.dtype, layout=torch.strided, 
+                                         device=x.device, pin_memory=False)
+                else:
+                    raise ValueError("Invalid num_inputs")
+                x_permuted = x[p]
+                sorted_tensor, _ = torch.sort(x_permuted)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_randperm.py:32]: Please add comment here on why do you sort

File: src/frontends/pytorch/src/op/randperm.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (vijaykr338) [reply src/frontends/pytorch/src/op/randperm.cpp]: I apologise, I re-evaluated the tests, now all of them pass including n == 0

[the remark this answers is not in the document]
Comment (vijaykr338) [reply src/frontends/pytorch/src/op/randperm.cpp]: also the implementation now handles n dynamically

File: tests/layer_tests/pytorch_tests/test_polar.py (hunk 9 above)
Comment (vijaykr338) [reply tests/layer_tests/pytorch_tests/test_polar.py]: ah yes, I have removed them, I was initially using atol and rtol as I was worried about the floating point differences that might be produced, but the default test layer handles it fine!

**Review by mvafin — APPROVED** [review]
Thanks for your contribution. Everything seems to be okay now.

**Comment by vijaykr338** [comment]
Hi @mvafin, are there any issues with the merging?

**Comment by mvafin** [comment]
> Hi @mvafin, are there any issues with the merging?

Don't worry about this. That is cause by the issues in our CI, I will look into it.

**Comment by mvafin** [comment]
build_jenkins
